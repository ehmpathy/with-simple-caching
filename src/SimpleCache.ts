export enum SimpleCacheExecutionMode {
  ASYNCHRONOUS = 'ASYNCHRONOUS',
  SYNCHRONOUS = 'SYNCHRONOUS',
}

/**
 * a simple cache which asynchronously gets and sets values to its store
 */
export interface SimpleAsyncCache<T> {
  meta: {
    execution: SimpleCacheExecutionMode.ASYNCHRONOUS;
  };
  get: (key: string) => Promise<T | undefined>;
  set: (key: string, value: T | undefined, options?: { secondsUntilExpiration?: number }) => Promise<void>;
}

/**
 * a simple cache which synchronously gets and sets values to its store
 */
export interface SimpleSyncCache<T> {
  meta: {
    execution: SimpleCacheExecutionMode.SYNCHRONOUS;
  };
  get: (key: string) => T | undefined;
  set: (key: string, value: T | undefined, options?: { secondsUntilExpiration?: number }) => void;
}

/**
 * a simple cache
 */
export type SimpleCache<T> = SimpleAsyncCache<T> | SimpleSyncCache<T>;

/**
 * a method which specifies where how to extract a simple-cache from input args
 */
export type SimpleCacheExtractionMethod<LI extends any[], C extends SimpleCache<any>> = (args: { fromInput: LI }) => C;
