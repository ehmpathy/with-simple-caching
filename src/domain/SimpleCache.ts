/**
 * a simple cache which synchronously gets and sets values to its store
 */
export interface SimpleSyncCache<T> {
  get: (key: string) => T | undefined;
  set: (key: string, value: T | undefined, options?: { secondsUntilExpiration?: number }) => void;
}

/**
 * a simple cache which asynchronously gets and sets values to its store
 */
export interface SimpleAsyncCache<T> {
  get: (key: string) => Promise<T | undefined>;
  set: (key: string, value: T | undefined, options?: { secondsUntilExpiration?: number }) => Promise<void>;
}

/**
 * a simple cache
 */
export type SimpleCache<T> = SimpleAsyncCache<T> | SimpleSyncCache<T>;

/**
 * a method which specifies where how to extract a simple-cache from input args
 */
export type SimpleCacheExtractionMethod<LI extends any[], C extends SimpleCache<any>> = (args: { fromInput: LI }) => C;
