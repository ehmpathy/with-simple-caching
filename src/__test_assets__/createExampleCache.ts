import { UniDuration } from '@ehmpathy/uni-time';

import { SimpleAsyncCache, SimpleSyncCache } from '../domain/SimpleCache';

export const createExampleSyncCache = () => {
  const store: Record<string, any> = {};
  const cache: SimpleSyncCache<any> = {
    set: (
      key: string,
      value: any,
      options?: { expiration?: UniDuration | null },
    ) => {
      store[key] = options ? { value, options } : { value };
    },
    get: (key: string) => store[key]?.value,
  };
  return { cache, store };
};

export const createExampleAsyncCache = <T>() => {
  const store: Record<string, { value: T; options?: any } | undefined> = {};
  const cache: SimpleAsyncCache<T> = {
    set: async (
      key: string,
      value: T | undefined,
      options?: { expiration?: UniDuration | null },
    ) => {
      // eslint-disable-next-line no-nested-ternary
      store[key] =
        value !== undefined
          ? options
            ? { value, options }
            : { value }
          : undefined;
    },
    get: async (key: string) => store[key]?.value,
  };
  return { cache, store };
};
