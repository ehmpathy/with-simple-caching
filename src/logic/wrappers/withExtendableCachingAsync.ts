import { isAFunction } from 'type-fns';
import { SimpleAsyncCache } from '../../domain/SimpleCache';
import { getCacheFromCacheOptionOrFromForKeyArgs } from '../options/getCacheFromCacheOptionOrFromForKeyArgs';
import { defaultKeySerializationMethod, defaultValueDeserializationMethod, defaultValueSerializationMethod } from '../serde/defaults';
import { withSimpleCachingAsync, WithSimpleCachingAsyncOptions } from './withSimpleCachingAsync';

/**
 * enumerates the extendable methods which can trigger cache operations
 */
export enum WithExtendableCachingTrigger {
  EXECUTE = 'EXECUTE',
  INVALIDATE = 'INVALIDATE',
  UPDATE = 'UPDATE',
}

/**
 * a simple typeguard which checks if an object has a property named `forInput`
 */
export const hasForInputProperty = (obj: any): obj is { forInput: any } => !!obj.forInput;

/**
 * the shape of logic that was wrapped with extendable caching for an async cache
 */
export interface LogicWithExtendableCachingAsync<
  /**
   * the logic we are caching the responses for
   */
  L extends (...args: any) => Promise<any>,
  /**
   * the type of cache being used
   */
  C extends SimpleAsyncCache<any>
> {
  /**
   * execute the logic with caching
   */
  execute: L;

  /**
   * invalidate the cached value for a given input
   *
   * note
   * - applies key serialization on the key input, just like execute
   */
  invalidate: (
    args:
      | {
          /**
           * invalidate the cache for this input
           */
          forInput: Parameters<L>;
        }
      | {
          /**
           * invalidate the cache for this key
           */
          forKey: string;

          /**
           * the cache to use, if the cache must be was defined from input parameters at runtime
           */
          cache?: C;
        },
  ) => Promise<void>;

  /**
   * update the cached value for a given input
   *
   * note
   * - applies key serialization on the key input, just like execute
   * - applies value serialization on the value output, just like execute
   */
  update: (
    args:
      | {
          /**
           * update the cache for this input
           */
          forInput: Parameters<L>;

          /**
           * update the cache to this value
           */
          toValue:
            | Awaited<ReturnType<L>>
            | ((args: { fromCachedOutput: Awaited<ReturnType<L>> | undefined }) => ReturnType<L> | Awaited<ReturnType<L>>);
        }
      | {
          /**
           * update the cache for this string
           */
          forKey: string;

          /**
           * update the cache to this value
           */
          toValue:
            | Awaited<ReturnType<L>>
            | ((args: { fromCachedOutput: Awaited<ReturnType<L>> | undefined }) => ReturnType<L> | Awaited<ReturnType<L>>);

          /**
           * the cache to use, if the cache must be was defined from input parameters at runtime
           */
          cache?: C;
        },
  ) => Promise<void>;
}

/**
 * exposes the cache-wrapped method along with some primitives which enable extending the caching logic
 *
 * specifically
 * - exposes a way to `invalidate` the cache, for a given input (e.g., to support external triggers for invalidation)
 * - exposes a way to `update` the cache, for a given input (e.g., to support write-through caching and optimistic caching)
 *
 * relevance
 * - when wrapping logic to cache the user is able to specify several caching options (e.g., key serialization method, value serialization method, etc)
 * - in order to define their own `invalidation` and `update` methods, without this function, the user would need to access these caching options per function elsewhere
 * - this function makes it easy to utilize and extend cache invalidation + update commands for the wrapped logic, by managing the references to the caching options on behalf of the user
 */
export const withExtendableCachingAsync = <
  /**
   * the logic we are caching the responses for
   */
  L extends (...args: any) => Promise<any>,
  /**
   * the type of cache being used
   */
  C extends SimpleAsyncCache<any>
>(
  logic: L,
  options: WithSimpleCachingAsyncOptions<L, C>,
): LogicWithExtendableCachingAsync<L, C> => {
  const execute: LogicWithExtendableCachingAsync<L, C>['execute'] = withSimpleCachingAsync(logic, options);

  const invalidate: LogicWithExtendableCachingAsync<L, C>['invalidate'] = async (args) => {
    // define how to get the cache, with support for `forKey` input instead of full input
    const cache = getCacheFromCacheOptionOrFromForKeyArgs({ args, options, trigger: WithExtendableCachingTrigger.INVALIDATE });

    // define the key, with support for `forKey` input instead of `forInput`
    const serializeKey = options.serialize?.key ?? defaultKeySerializationMethod;
    const key = hasForInputProperty(args) ? serializeKey({ forInput: args.forInput }) : args.forKey;

    // set undefined into the cache for this key, to invalidate the cached value
    await cache.set(key, undefined);
  };

  const update: LogicWithExtendableCachingAsync<L, C>['update'] = async (args) => {
    // define how to get the cache, with support for `forKey` input instead of full input
    const cache = getCacheFromCacheOptionOrFromForKeyArgs({ args, options, trigger: WithExtendableCachingTrigger.UPDATE });

    // define the key, with support for `forKey` input instead of `forInput`
    const serializeKey = options.serialize?.key ?? defaultKeySerializationMethod;
    const key = hasForInputProperty(args) ? serializeKey({ forInput: args.forInput }) : args.forKey;

    // deserialize the cached value
    const cachedValue = await cache.get(key);
    const deserializeValue = options.deserialize?.value ?? defaultValueDeserializationMethod;
    const deserializedCachedOutput = cachedValue !== undefined ? deserializeValue(cachedValue) : undefined;

    // compute the new value
    const newValue: Awaited<ReturnType<L>> = isAFunction(args.toValue)
      ? await args.toValue({ fromCachedOutput: deserializedCachedOutput })
      : await args.toValue;

    // define the serialized new value
    const serializeValue = options.serialize?.value ?? defaultValueSerializationMethod;
    const serializedNewValue = serializeValue(newValue);

    // set the new value for this key
    await cache.set(key, serializedNewValue, { secondsUntilExpiration: options.secondsUntilExpiration });
  };

  return { execute, invalidate, update };
};
