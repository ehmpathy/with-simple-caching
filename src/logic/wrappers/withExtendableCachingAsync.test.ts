import { createExampleAsyncCache } from '../../__test_assets__/createExampleCache';
import { SimpleAsyncCache } from '../../domain/SimpleCache';
import { BadRequestError } from '../../utils/errors/BadRequestError';
import { withExtendableCachingAsync } from './withExtendableCachingAsync';

describe('withExtendableCachingAsync', () => {
  describe('execute', () => {
    it('should be able to execute logic', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async () => {
          apiCalls.push(1);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute();
      const result2 = await callApi.execute();
      const result3 = await callApi.execute();

      // check that the response is the same each time
      expect(result1).toEqual(result2);
      expect(result2).toEqual(result3);

      // check that "api" was only called once
      expect(apiCalls.length).toEqual(1);
    });
  });
  describe('invalidate', () => {
    it('should be able to invalidate a cached value by input', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async ({ galaxy }: { galaxy: string }) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' });
      const result2 = await callApi.execute({ galaxy: 'pegasus' });
      const result3 = await callApi.execute({ galaxy: 'andromeda' });
      const result4 = await callApi.execute({ galaxy: 'pegasus' });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // invalidate the cached value for one of the inputs
      await callApi.invalidate({ forInput: [{ galaxy: 'andromeda' }] });

      // now call the cache again for that invalidated value, and prove it called the api again
      const result5 = await callApi.execute({ galaxy: 'andromeda' });
      expect(result5).not.toEqual(result1);
      expect(apiCalls.length).toEqual(3);
    });
    it('should be able to invalidate a cached value by key', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async ({ galaxy }: { galaxy: string }) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' });
      const result2 = await callApi.execute({ galaxy: 'pegasus' });
      const result3 = await callApi.execute({ galaxy: 'andromeda' });
      const result4 = await callApi.execute({ galaxy: 'pegasus' });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // invalidate the cached value for one of the inputs
      await callApi.invalidate({
        forKey: JSON.stringify([{ galaxy: 'andromeda' }]),
      });

      // now call the cache again for that invalidated value, and prove it called the api again
      const result5 = await callApi.execute({ galaxy: 'andromeda' });
      expect(result5).not.toEqual(result1);
      expect(apiCalls.length).toEqual(3);
    });
    it('should be able to invalidate a cached value by key when the cache is to be defined at runtime from inputs', async () => {
      // define an example fn
      const { cache } = createExampleAsyncCache<string>();
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async (
          { galaxy }: { galaxy: string },
          _: { cache: SimpleAsyncCache<string> },
        ) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        {
          cache: ({ fromInput }) => fromInput[1].cache,
          serialize: {
            key: ({ forInput }) => forInput[0].galaxy, // dont include cache as part of the key + simplify the key to just the galaxy
          },
        },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' }, { cache });
      const result2 = await callApi.execute({ galaxy: 'pegasus' }, { cache });
      const result3 = await callApi.execute({ galaxy: 'andromeda' }, { cache });
      const result4 = await callApi.execute({ galaxy: 'pegasus' }, { cache });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // prove that it will throw a helpful error if we dont explicitly pass in the cache in this case
      try {
        await callApi.invalidate({ forKey: 'andromeda' });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error).toBeInstanceOf(BadRequestError);
        if (!(error instanceof BadRequestError))
          throw new Error('error should have been instance of BadRequestError'); // satisfy typescript defs
        expect(error.message).toContain(
          'could not find the cache to invalidate',
        );
        expect(error.message).toMatchSnapshot();
      }

      // invalidate the cached value for one of the inputs
      await callApi.invalidate({ forKey: 'andromeda', cache });

      // now call the cache again for that invalidated value, and prove it called the api again
      const result5 = await callApi.execute({ galaxy: 'andromeda' }, { cache });
      expect(result5).not.toEqual(result1);
      expect(apiCalls.length).toEqual(3);
    });
  });
  describe('update', () => {
    it('should be able to update a cached value by input', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async ({ galaxy }: { galaxy: string }) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' });
      const result2 = await callApi.execute({ galaxy: 'pegasus' });
      const result3 = await callApi.execute({ galaxy: 'andromeda' });
      const result4 = await callApi.execute({ galaxy: 'pegasus' });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // update the cached value for one of the inputs
      await callApi.update({
        forInput: [{ galaxy: 'andromeda' }],
        toValue: 821,
      });

      // now call the cache again
      const result5 = await callApi.execute({ galaxy: 'andromeda' });

      // and prove that the value changed to what we wanted to update it to
      expect(result5).toEqual(821);
      expect(result5).not.toEqual(result1);

      // and prove that we didn't call the api again
      expect(apiCalls.length).toEqual(2);
    });
    it('should be able to update a cached value by key', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async ({ galaxy }: { galaxy: string }) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' });
      const result2 = await callApi.execute({ galaxy: 'pegasus' });
      const result3 = await callApi.execute({ galaxy: 'andromeda' });
      const result4 = await callApi.execute({ galaxy: 'pegasus' });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // update the cached value for one of the inputs
      await callApi.update({
        forKey: JSON.stringify([{ galaxy: 'andromeda' }]),
        toValue: 821,
      });

      // now call the cache again
      const result5 = await callApi.execute({ galaxy: 'andromeda' });

      // and prove that the value changed to what we wanted to update it to
      expect(result5).toEqual(821);
      expect(result5).not.toEqual(result1);

      // and prove that we didn't call the api again
      expect(apiCalls.length).toEqual(2);
    });
    it('should be able to update a cached value by key when the cache is to be defined at runtime from inputs', async () => {
      // define an example fn
      const { cache } = createExampleAsyncCache<string>();
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async (
          { galaxy }: { galaxy: string },
          _: { cache: SimpleAsyncCache<string> },
        ) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        {
          cache: ({ fromInput }) => fromInput[1].cache,
          serialize: {
            key: ({ forInput }) => forInput[0].galaxy, // dont include cache as part of the key + simplify the key to just the galaxy
          },
        },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' }, { cache });
      const result2 = await callApi.execute({ galaxy: 'pegasus' }, { cache });
      const result3 = await callApi.execute({ galaxy: 'andromeda' }, { cache });
      const result4 = await callApi.execute({ galaxy: 'pegasus' }, { cache });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // prove that it will throw a helpful error if we dont explicitly pass in the cache in this case
      try {
        await callApi.update({ forKey: 'andromeda', toValue: 821 });
        throw new Error('should not reach here');
      } catch (error) {
        expect(error).toBeInstanceOf(BadRequestError);
        if (!(error instanceof BadRequestError))
          throw new Error('error should have been instance of BadRequestError'); // satisfy typescript defs
        expect(error.message).toContain('could not find the cache to update');
        expect(error.message).toMatchSnapshot();
      }

      // invalidate the cached value for one of the inputs
      await callApi.update({ forKey: 'andromeda', toValue: 821, cache });

      // now call the cache again
      const result5 = await callApi.execute({ galaxy: 'andromeda' }, { cache });

      // and prove that the value changed to what we wanted to update it to
      expect(result5).toEqual(821);
      expect(result5).not.toEqual(result1);

      // and prove that we didn't call the api again
      expect(apiCalls.length).toEqual(2);
    });
    it('should be able to update a cached value by input to a value based on prior value', async () => {
      // define an example fn
      const apiCalls = [];
      const callApi = withExtendableCachingAsync(
        async ({ galaxy }: { galaxy: string }) => {
          apiCalls.push(galaxy);
          return Math.random();
        },
        { cache: createExampleAsyncCache().cache },
      );

      // call the fn a few times
      const result1 = await callApi.execute({ galaxy: 'andromeda' });
      const result2 = await callApi.execute({ galaxy: 'pegasus' });
      const result3 = await callApi.execute({ galaxy: 'andromeda' });
      const result4 = await callApi.execute({ galaxy: 'pegasus' });

      // check that the response is the same each time the input is the same
      expect(result1).toEqual(result3);
      expect(result2).toEqual(result4);

      // check that "api" was only called twice
      expect(apiCalls.length).toEqual(2);

      // update the cached value for one of the inputs
      await callApi.update({
        forInput: [{ galaxy: 'andromeda' }],
        toValue: ({ fromCachedOutput }) => (fromCachedOutput ?? 0) * 2,
      });

      // now call the cache again
      const result5 = await callApi.execute({ galaxy: 'andromeda' });

      // and prove that the value changed to what we wanted to update it to
      expect(result5).not.toEqual(result1);
      expect(result5).toEqual(result1 * 2);

      // and prove that we didn't call the api again
      expect(apiCalls.length).toEqual(2);
    });
  });
});
