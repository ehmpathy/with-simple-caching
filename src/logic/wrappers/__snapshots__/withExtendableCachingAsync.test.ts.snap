// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withExtendableCachingAsync invalidate should be able to invalidate a cached value by key when the cache is to be defined at runtime from inputs 1`] = `
"BadRequestError: could not find the cache to invalidate in. invalidate was called forKey but the cache for this method was defined as a function which retrieves the cache from the input. therefore, since there is no input accessible, the cache should have been explicitly passed in on invalidate

{\\"args\\":{\\"forKey\\":\\"andromeda\\"}}}"
`;

exports[`withExtendableCachingAsync update should be able to update a cached value by key when the cache is to be defined at runtime from inputs 1`] = `
"BadRequestError: could not find the cache to update in. update was called forKey but the cache for this method was defined as a function which retrieves the cache from the input. therefore, since there is no input accessible, the cache should have been explicitly passed in on update

{\\"args\\":{\\"forKey\\":\\"andromeda\\",\\"toValue\\":821}}}"
`;
